### DISCUSSION TITLE
DevMeeting - 2021-07-26/27

### DISCUSSION THREAD

**Discussion by jackkoenig** [chipsalliance/chisel#2055] [Dev Meetings Recording]
### Outreach

* [jiuyang] CCC Shanghai
    * Done, need Jack to make it public.
* [jiuyang] None resolved Q&A
    * Interesting hardware algorithm question on SO: [https://stackoverflow.com/questions/68476482/what-is-a-low-latency-hardware-algorithm-to-find-the-maximum-value-in-an-array](https://stackoverflow.com/questions/68476482/what-is-a-low-latency-hardware-algorithm-to-find-the-maximum-value-in-an-array)
        * [https://github.com/sequencer/arithmetic](https://github.com/sequencer/arithmetic)	
* [jiuyang] Send calendar invites on the mailing list?
    * Added [Chisel Dev Meeting_c_hisamsd8afoqha1m7032qdpfig@group.calendar.google.com.ics](https://calendar.google.com/calendar/u/0/exporticalzip?cexp=Y19oaXNhbXNkOGFmb3FoYTFtNzAzMnFkcGZpZ0Bncm91cC5jYWxlbmRhci5nb29nbGUuY29t)
    * Trying to figure out how to let others modify the event

### Pinned Items

* Priority PRs:
    * [jiuyang] api-config-chipsalliance publishing
        * [jiuyang] PRs:
            * [https://github.com/chipsalliance/api-config-chipsalliance/pull/14](https://github.com/chipsalliance/api-config-chipsalliance/pull/14)
            * [https://github.com/chipsalliance/api-config-chipsalliance/pull/15](https://github.com/chipsalliance/api-config-chipsalliance/pull/15)
            * [https://github.com/chipsalliance/api-config-chipsalliance/pull/16](https://github.com/chipsalliance/api-config-chipsalliance/pull/16)
            * [https://github.com/chipsalliance/api-config-chipsalliance/pull/17](https://github.com/chipsalliance/api-config-chipsalliance/pull/17)
            * [https://github.com/chipsalliance/api-config-chipsalliance/pull/18](https://github.com/chipsalliance/api-config-chipsalliance/pull/18)
        * [jiuyang] New bug found when writing the documentation.
        * [jiuyang] available via `me.jiuyang.cde::cde:0.1` now, when those PRs merged, I’ll remove my fork.
    * [chick] [https://github.com/ucb-art/rocket-dsp-utils/issues/30](https://github.com/ucb-art/rocket-dsp-utils/issues/30)
    * [jiuyang] [https://github.com/chipsalliance/firrtl/pull/2298](https://github.com/chipsalliance/firrtl/pull/2298) 

### Discussion 1 

* [schuyler] Constant Propagation on Output Ports used as RHS Values
    * [https://github.com/chipsalliance/firrtl/issues/2299](https://github.com/chipsalliance/firrtl/issues/2299) 
    * We could probably do the constprop optimization in SFC
    * Better, we should canonicalize away lvalues used as rvalues
        * [kevin] canonicalizing ports to SSA could simplify other passes
        * See approach for SMT backend: [https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/backends/experimental/smt/FirrtlToTransitionSystem.scala#L612](https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/backends/experimental/smt/FirrtlToTransitionSystem.scala#L612)
            * Worth integrating into a standard transform?
    * For the FEV problem, rather than trying to keep optimizations in a sync, a better approach might be to record transformations that do not maintain formal equivalence in order to generate waivers/constraints to the FEV tool
    * [kevin] You can dontTouch the registers when there’s an FEV mismatch
        * Good for bugfinding but not for CI
* [jack] Anyone interested can see DataView “pre-pub” docs: [https://github.com/chipsalliance/chisel3/pull/1955](https://github.com/chipsalliance/chisel3/pull/1955)
* [jiuyang] Can we get in touch with original original espresso author(s) about the license?

### Discussion 2

* [juyang] Should we mandate the chisel3 plugin in 3.5?
    * 3 options
        * Don’t ever mandate plugin
        * Deprecate no plugin in 3.5
        * Mandate plugin in 3.5
            * We could deprecate no plugin in 3.4.4, but also make sure the error message when plugin is mandated is clear and points to docs
    * Need to resolve (From July 12th meeting)
        * [https://github.com/chipsalliance/chisel3/issues/1890](https://github.com/chipsalliance/chisel3/issues/1890)
            * Plugin doesn’t do prefixing for non-Data subtypes but chiselName does
            * [jack] If resolution is “Won’t Fix”, then I think we should add “prefix” macro
                * val foo = prefix(“foo”) { … }
                * Could do ValName to have: val foo = prefix { … }
            * Another option is to add a “Prefixable” trait that user-defined types can implement
                * In @chiselName we had “DoNotPrefix”
                * Note that this means you can’t prefix Scala stdlib types (Tuple, List, etc.)
                    * Could do a typeclass?
        * [https://github.com/chipsalliance/chisel3/issues/1893](https://github.com/chipsalliance/chisel3/issues/1893)
        * [https://github.com/chipsalliance/chisel3/issues/1894](https://github.com/chipsalliance/chisel3/issues/1894)
* [kevin] Can we make it a rule not to merge PRs that add deprecations without removing all uses of the class/method/etc in our code base?
    * if we leave in deprecated code, we create technical dept
    * if we don’t “dogfood” our own deprecation it can also hide bugs like: [https://github.com/chipsalliance/firrtl/pull/2303#issuecomment-887678669](https://github.com/chipsalliance/firrtl/pull/2303#issuecomment-887678669)
    * [jack] Yes we should do this, there are cases where it’s not possible but those should require a warning ignore callout
* [adam] Can we actually remove the API when we deprecate
    * The PR to master is the removal, the backport is the deprecation
    * This does mean deprecations will come in minor versions rather than .0
    * This also could leverage Mima to make sure we do the deprecation right (ie. the removal would fail)
    * Massive advantage of this is that by deleting the API, you *must* update the tests so you will see if there is something missing by deleting the API, helps answer the question “can we delete this API? Do we have a replacement?”
    * [jack] Little tweak is that we could have these deprecations show up in .0 if we start long-term branches for “next major version” before releasing .0
        * Eg. 3.5 could already be a branch that we backport to
        * Resolution: Kept under advisory, but allowing deprecations on minor versions will be very helpful for our code quality (hacks to maintain bad APIs won’t end up on master anymore as often)
    * **Important Resolution: Deprecations on minor versions will happen more frequently to facilitate better development flow and more agility**
        * Users should always bump to the latest minor version of the current major version before bumping to the next major version to ensure they are seeing all deprecation warnings

**Comment by jackkoenig** [comment]
A longer form description of the update to deprecations development practice (originally written https://github.com/chipsalliance/chisel3/pull/2052#discussion_r677727019)

We discussed a change to deprecation policy that deprecations on minor versions are now the standard with removal in the subsequent major version. Our reasoning is that this enables better development practice, cleaner code, and more agility.

A basic problem we've had for a long time is that we often deprecate and never remove, and sometimes even deprecate things without replacement. The new development flow is that, if an API is to be changed, it is deleted on master and backported to the latest major version. This deletion on master will keep the main codebase cleaner, while also requiring the developer to update all tests thus ensuring we know what the code migration path looks like and that it is appropriate to deprecate the API.

The backport will then fail (because it won't be binary compatible), and the developer is then required to deprecate the API on the backport. This has a massive benefit that the internal hacks that are often necessary to maintain some deprecated APIs will only have to be made on the backport branch and will never exist on master.

Finally, it's true that people may sometimes skip minor versions. It will become part of our release notes and announcement notes which minor version users need to bump to before bumping to the major release. This requires a bit of additional discipline but I believe the benefits to be worth it.

**Reply by mwachs5** [reply]
This sounds great... should we document this in the "what version should I use" section of the docs?

**Reply by jackkoenig** [reply]
Yes absolutely
